pkg/clusterapi/aws/utils.go

```diff
@@ -0,0 +1,157 @@
+// slice after sorting will be the most recently launched, and can be considered the definitive instance
+// for the machine, a caller may wish to terminate all others. This function should only be called with
+// running instances, not those which are stopped or terminated.
+func SortInstances(instances []*ec2.Instance) {
+	sort.Slice(instances, func(i, j int) bool {
+		if instances[i].LaunchTime == nil && instances[j].LaunchTime == nil {
+			// No idea what to do here, should not be possible, just return the first.
+			return false
+		}
+		if instances[i].LaunchTime != nil && instances[j].LaunchTime == nil {
+			return true
+		}
+		if instances[i].LaunchTime == nil && instances[j].LaunchTime != nil {
+			return false
+		}
+		if (*instances[i].LaunchTime).After(*instances[j].LaunchTime) {
+			return true
+		}
+		return false
+	})
+}
+
+// GetInstance returns the AWS instance for a given machine. If multiple instances match our machine,
+// the most recently launched will be returned. If no instance exists, the exists bool will return false.
+func GetInstance(machine *clusterv1.Machine, client ec2iface.EC2API) (*ec2.Instance, bool, error) {
+	instances, err := GetRunningInstances(machine, client)
+	if err != nil {
+		return nil, false, err
+	}
+	if len(instances) == 0 {
+		return nil, false, nil
+	}
+
+	SortInstances(instances)
+	return instances[0], true, nil
+}
+
+// GetRunningInstances returns all running instances that have a tag matching our machine name,
+// and cluster ID.
+func GetRunningInstances(machine *clusterv1.Machine, client ec2iface.EC2API) ([]*ec2.Instance, error) {
+	instances := []*ec2.Instance{}
+
+	machineName := machine.Name
+
+	clusterID, ok := getClusterID(machine)
+	if !ok {
+		return instances, fmt.Errorf("unable to get cluster ID for machine: %s", machine.Name)
+	}
+
+	// Query instances with our machine's name, and in running/pending state.
+	request := &ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("tag:Name"),
+				Values: []*string{&machineName},
+			},
+			{
+				Name:   aws.String("instance-state-name"),
+				Values: []*string{aws.String("running"), aws.String("pending")},
+			},
+			{
+				Name:   aws.String("tag:clusterid"),
+				Values: []*string{&clusterID},
+			},
+		},
+	}
+	result, err := client.DescribeInstances(request)
+	if err != nil {
+		return instances, err
+	}
+
+	for _, reservation := range result.Reservations {
+		for _, instance := range reservation.Instances {
+			instances = append(instances, instance)
+		}
+	}
+
+	return instances, nil
+}
+
+// CreateAWSClients creates clients for the AWS services we use, using either the cluster AWS credentials secret
+// if defined (i.e. in the root cluster), otherwise the IAM profile of the master where the
+// actuator will run. (target clusters)
+func CreateAWSClients(kubeClient kubernetes.Interface, mSpec *cov1.MachineSetSpec, namespace, region string) (ec2iface.EC2API, elbiface.ELBAPI, error) {
```

Comment:
Let's have this create and return a struct that encapsulates the various AWS clients. Let's also create an interface for the other functions that the struct implements. Then, we can unit test the actuator with a mock.